main/static/script.js

```diff
@@ -94,107 +74,98 @@ module.controller("mainCtrl", function(
-    $scope.inputModel.renderType = queries["renderType"] || "line";
-    $scope.inputModel.profile = parseBool(queries["profile"]);
-    if ($scope.inputModel.query) {
-      launchRequest({
-        profile: $scope.inputModel.profile,
-        query:   $scope.inputModel.query
-      });
-    }
-  });
-
-  // true if the output should be tabular.
-  $scope.isTabular = function() {
-    return ["describe all", "describe metrics", "describe"].indexOf($scope.queryResult.name) >= 0;
+  this.dec = function() {
+    count--;
   };
-  $scope.hasProfiling = function() {
-    return !!($scope.queryResult && $scope.queryResult.profile);
+  this.pos = function() {
+    return count > 0;
+  };
+  this.zero = function() {
+    return count === 0;
   };
+};
+// A singleton Counter for launched queries.
+module.factory("$launchedQueries", function() {
+  return new Counter();
+});
+// A singleton Counter for waiting charts
+module.factory("$chartWaiting", function() {
+  return new Counter();
+});
 
-  $scope.screenState = function() {
-    if ($scope.launchedQuery > 0) {
-      return "loading";
-    } else if ($scope.launchedQuery === 0 && $scope.chartWaiting > 0) {
-      return "rendering";
-    } else if ($scope.queryResult && !$scope.queryResult.success) {
-      return "error";
-    } else {
-      return "rendered";
+// A ticket booth will give you a ticket with .next()
+// The ticket will be .valid() until another ticket has been asked for.
+function TicketBooth() {
+  var count = 0;
+  function Ticket(n) {
+    this.valid = function() {
+      return n === count;
     }
+  }
+  this.next = function() {
+    return new Ticket(++count);
   };
+}
+// A singleton ticketbooth for query counting
+module.factory("$queryTicketBooth", function() {
+  return new TicketBooth();
+});
 
-  $scope.chartWaiting = 0;
-  $scope.launchedQuery = 0;
+// chart-related functions
+function clearChart(chart) {
+  if (chart.chart !== null) {
+    chart.chart.clearChart();
+    chart.chart = null;
+  }
+};
+
+module.factory("$asyncRender", function($chartWaiting) {
+  return function($scope, chart, dataTable, options) {
+    google.visualization.events.addListener(chart.chart, "ready", function() {
+      $scope.$apply(function() { $chartWaiting.dec(); });
+    });
+    setTimeout(function(){chart.chart.draw(dataTable, options)}, 1);
+  };
+});
 
-  function launchRequest(params) {
-    queryCounter++;
-    var currentQueryCounter = queryCounter; // store it in the closure.
-    $scope.launchedQuery++;
+module.factory("$launchRequest", function($google, $http, $receive, $queryTicketBooth, $launchedQueries) {
+  return function($scope, params) {
+    var ticket = $queryTicketBooth.next();
+    $launchedQueries.inc();
     var request = $http.get("/query", {
       params:params
     }).success(function(data, status, headers, config) {
-      $scope.launchedQuery--;
+      $launchedQueries.dec();
       $scope.queryResult = data;
```

Comment:
I used a `$q.defer()` promise to clean it up! Now the callers just do `.then(function(data) { $scope.queryResult = data; });`